### ISSUE TITLE
Allow generation of JavaScript

### ISSUE THREAD

**Issue by sigkell** [atom/package-generator#14]
It would be nice to have the option to generate JavaScript/JSON over CoffeeScript/CSON. For those who dislike CoffeeScript (such as myself), it's a little annoying.

It appears you'd need to add this functionality to `apm init` however, before you could create a UI for selection.

**Comment by frozzare** [comment]
+1 for JavaScript!

**Comment by Trudko** [comment]
+1

**Comment by thedaniel** [comment]
ES6 classes are compatible with the Coffeescript classes used in Atom, so this is more plausible now that Atom supports `"use babel";`.

**Comment by keoghpe** [comment]
+1

**Comment by amiuhle** [comment]
:+1: 
`'use babel;`

**Comment by davegomez** [comment]
+1... And make it ES6 ready too =)

**Comment by saitodisse** [comment]
:+1:

**Comment by thedaniel** [comment]
No need to comment with :+1: unless you have some new information to add to the issue.

**Comment by stevenhauser** [comment]
Now that https://github.com/atom/apm/issues/417 has been merged, this should be pretty trivial to add. Thoughts on implementation? I see three main options:

1. Add a separate command for JS and CS packages. This would split `package-generator:generate-package` into `package-generator:generate-javascript-package` and `package-generator:generate-coffeescript-package` (or something similar).
2. Add a secondary modal after choosing `package-generator:generate-package` to choose between JS and CS.
3. Add a setting for which language you want to generate packages in.

My vote would be for 3. I think most devs will generate packages in *either* JS or CS but probably not switch between both (at least when creating *new* packages). Number 2 seems kind of annoying and makes the flow different b/w package and theme generation, though I suppose that's not a huge deal. Number 1 seems like you could mistakenly choose CS or JS which would be irritating.

Thoughts anyone? Different idea?

**Comment by thedaniel** [comment]
I'm down with a sticky setting and a default to javascript, second choice would be a sticky setting but a modal on first run that asks your preference. @atom/feedback any thoughts?

**Comment by varemenos** [comment]
Imho, the setting shouldn't stick.
The user should be asked every time whether he wants JS or CS when they try to create a new package.

**Comment by stevenhauser** [comment]
Note that the sticky setting has been implemented in #36.

@thedaniel what do you see as the value of a modal on first run? Discoverability? Also I'm unsure how that first-run-only would be implemented—is there some standard way to store that state in Atom?

@varemenos Why do you think it should ask every time? I see very few people—if any—wanting to develop *new* packages in different languages. Is that something you'd use?

**Comment by lee-dohm** [comment]
Personally, I think this would be best:

1. Modal asking for preference if `package-generator.generateLanguage` is unset
1. Modal asks for preference with "Don't ask me again" checkbox
1. If preference is selected with "Don't ask me again" *unchecked*, it generates with that language and **does not** store the preference in the config
1. If preference is selected with "Don't ask me again" *checked*, it generates with that language and **does** store the preference in the config

**Comment by stevenhauser** [comment]
@lee-dohm to me that's adding a lot of UI complexity when we don't know if anyone would actually benefit from all of those options. Also do you have an example of a package that creates a list modal with a checkbox in it?

Again, I'm really curious if anyone has a real use case for frequently developing packages in different languages? It just seems really unlikely to me which is why I was inclined keep the UI and code as simple as possible in #36—until people specifically request that feature be added.

**Comment by steelbrain** [comment]
Bump

**Comment by stevenhauser** [comment]
Hey it's been a few months since I opened #36 to address this issue—is there still interest in this? There's some conversation above about different approaches, but I think it'd be nice to go the simple route in #36 to at least have _something_ available, and then we could enhance it later.

**Comment by lee-dohm** [comment]
@stevenhauser I've added your PR to the queue to be reviewed. I believe there still is interest in it. There's just a lot of things in the queue :frowning: Bear with us please :grinning:

**Comment by stevenhauser** [comment]
@lee-dohm Thanks for adding it to the queue; I completely understand that you're all busy. I'll try to iron out any merge conflicts before it gets reviewed.

**Comment by lee-dohm** [comment]
Thanks very much!

**Comment by joshaber** [comment]
#36 has landed. :boom:

**Comment by Romaine** [comment]
Hi, I'm new to creating packages for atom.  I think @lee-dohm 's idea is the best way to go about this.  I dont know how to quote on this thing.  It seems the impracticality of having javascript as default was overlooked.  A user preference should be a user's preference and not a default setting.  I mean it does seem that the the intention is for packages to be written in coffee.  A user shouldn't have to manually dig around to a packages hidden settings page to have code generated in an ecosystems preferred language and encouraged language.  Not least have to dig up a closed issue ^^

tl;dr, in package.json, line 33 should say  "default": "coffeescript",

I'm not sure how things work so I assume this isn't worth a pull request.

**Comment by thomasjo** [comment]
> I mean it does seem that the the intention is for packages to be written in coffee

@Romaine That _used to be_ the intention. We are slowly migrating core code (and thus the default) towards JavaScript.

**Comment by Romaine** [comment]
Oh, is there a reason for that?  Some sort of blog post or mailing list or something that details best practices in that regard?  I mean who says?  It'd leave me with a better understanding of what language I should be putting time into learning properly.  I mean I don't wanna end up court short.

Link please?

Am I asking of too much; does that rule only apply to core code?
